examples/shor.py

```diff
@@ -0,0 +1,160 @@
+steps. In the first step, Quantum Phase Estimation is applied to a unitary such
+as
+
+    U|y⟩ = |xy mod n⟩  0 <= y < n
+    U|y⟩ = |y⟩         n =< y
+
+whose eigenvalues are s/r for s = 0, 1, ..., r - 1. In the second step, the
+classical continued fractions algorithm is used to recover r from s/r. Note
+that when gcd(s, r) > 1 then an incorrect r is found. This can be detected
+by verifying that r is indeed the order of x. If it is not, Quantum Phase
+Estimation algorithm is retried.
+
+[1]: https://arxiv.org/abs/quant-ph/9508027
+"""
+
+import argparse
+import math
+import random
+
+from typing import Callable, Optional
+
+parser = argparse.ArgumentParser(description='Factorization demo.')
+parser.add_argument('n', type=int, help='composite integer to factor')
+
+
+def naive_order_finder(x: int, n: int) -> int:
+    """Computes smallest positive r such that x^r mod n == 1.
+
+    Args:
+        x: integer whose order is to be computed, must be greater than one
+           and belong to the multiplicative group of integers modulo n (which
+           consists of positive integers relatively prime to n),
+        n: modulus of the multiplicative group.
+
+    Returns:
+        Smallest positive integer r such that x^r == 1 mod n.
+
+    Raises:
+        ValueError when x is 1 or not an element of the multiplicative
+        group of integers modulo n.
+    """
+    if x < 2 or n <= x or math.gcd(x, n) > 1:
+        raise ValueError(f'Invalid x={x} for modulus n={n}.')
+    r, y = 1, x
+    while y != 1:
+        y = (x * y) % n
+        r += 1
+    return r
+
+
+def quantum_order_finder(x: int, n: int) -> int:
+    """Computes smallest positive r such that x^r mod n == 1."""
+    raise NotImplementedError('Quantum order finder is not implemented yet.')
+
+
+def find_factor_of_prime_power(n: int) -> Optional[int]:
+    """Returns non-trivial factor of n if n is a prime power, else None."""
+    for k in range(2, math.floor(math.log2(n)) + 1):
+        c = math.pow(n, 1 / k)
+        c1 = math.floor(c)
+        if c1**k == n:
+            return c1
+        c2 = math.ceil(c)
+        if c2**k == n:
+            return c2
+    return None
+
+
+def find_factor(n: int,
+                order_finder: Callable[[int, int], int],
+                max_attempts: int = 30) -> Optional[int]:
+    """Returns a non-trivial factor of composite integer n.
+
+    Args:
+        n: integer to factorize,
+        order_finder: function for finding the order of elements of the
+            multiplicative group of integers modulo n,
+        max_attempts: number of random x's to try, also an upper limit
+            on the number of order_finder invocations.
+
+    Returns:
+        Non-trivial factor of n or None if no such factor was found.
+        Factor k of n is trivial if it is 1 or n.
+    """
+    if n == 2:
+        return None
+    if n % 2 == 0:
+        return 2
+    c = find_factor_of_prime_power(n)
```

Comment:
Use `sympy.primetest` https://docs.sympy.org/latest/_modules/sympy/ntheory/primetest.html to avoid the infinite loop.

Reply:
Done.